File: core/src/main/java/fr/sncf/osrd/stdcm/graph/STDCMEdge.java
package fr.sncf.osrd.stdcm.graph;

import fr.sncf.osrd.envelope.Envelope;
import fr.sncf.osrd.infra.api.signaling.SignalingRoute;
import fr.sncf.osrd.utils.graph.Pathfinding;
import java.util.Objects;

public record STDCMEdge(
        // Route considered for this edge
        SignalingRoute route,
        // Envelope of the train going through the route (starts at t=0). Does not account for allowances.
        Envelope envelope,
        // Time at which the train enters the route
        double timeStart,
        // Maximum delay we can add after this route by delaying the start time without causing conflicts
        double maximumAddedDelayAfter,
        // Delay we needed to add in this route to avoid conflicts (by shifting the departure time)
        double addedDelay,
        // Time of the next occupancy of the route, used to identify the "opening" used by the edge
        double timeNextOccupancy,
        // Total delay we have added by shifting the departure time since the start of the path
        double totalDepartureTimeShift,
        // Node located at the start of this edge, null if this is the first edge
        STDCMNode previousNode,
        // Offset of the envelope if it doesn't start at the beginning of the edge
        double envelopeStartOffset,
        // Time at which the train enters the route, discretized by only considering the minutes.
        // Used to identify visited edges
        int minuteTimeStart,
        // Speed factor used to account for standard allowance.
        // e.g. if we have a 5% standard allowance, this value is 1/1.05
        double standardAllowanceSpeedFactor,
        // Index of the last waypoint passed by this train
        int waypointIndex,
        // True if the edge end is a stop
        boolean endAtStop
) {
    @Override
    public boolean equals(Object other) {
        if (other == null || other.getClass() != STDCMEdge.class)
            return false;
        var otherEdge = (STDCMEdge) other;
        if (!route.getInfraRoute().getID().equals(otherEdge.route.getInfraRoute().getID()))
            return false;

        // We need to consider that the edges aren't equal if the times are different,
        // but if we do it "naively" we end up visiting the same places a near-infinite number of times.
        // We handle it by discretizing the start time of the edge: we round the time down to the minute and compare
        // this value.
        return minuteTimeStart == otherEdge.minuteTimeStart;
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                route.getInfraRoute().getID(),
                timeNextOccupancy
        );
    }

    /** Returns the node at the end of this edge */
    STDCMNode getEdgeEnd(STDCMGraph graph) {
        if (!endAtStop) {
            // We move on to the next route
            return new STDCMNode(
                    getTotalTime() + timeStart(),
                    envelope().getEndSpeed(),
                    graph.infra.getSignalingRouteGraph().incidentNodes(route()).nodeV(),
                    totalDepartureTimeShift(),
                    maximumAddedDelayAfter(),
                    this,
                    waypointIndex,
                    null,
                    -1
            );
        } else {
            // New edge on the same route, after a stop
            double stopDuration = graph.steps.get(waypointIndex + 1).duration();
            var newWaypointIndex = waypointIndex + 1;
            while (newWaypointIndex + 1 < graph.steps.size() && !graph.steps.get(newWaypointIndex + 1).stop())
                newWaypointIndex++; // Skip waypoints where we don't stop (not handled here)
            return new STDCMNode(
                    getTotalTime() + timeStart() + stopDuration,
                    0,
                    null,
                    totalDepartureTimeShift(),
                    maximumAddedDelayAfter(),
                    this,
                    newWaypointIndex,
                    new Pathfinding.EdgeLocation<>(route, envelopeStartOffset + getLength()),
                    stopDuration
            );
        }
    }

    /** Returns how long it takes to go from the start to the end of the route, accounting standard allowance. */
    public double getTotalTime() {
        return envelope.getTotalTime() / standardAllowanceSpeedFactor;
    }

    /** Returns the length of the edge */
    public double getLength() {
        return envelope.getEndPos();
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMEdge.kt
package fr.sncf.osrd.stdcm.graph

import fr.sncf.osrd.envelope.Envelope
import fr.sncf.osrd.sim_infra.api.Block
import fr.sncf.osrd.stdcm.infra_exploration.InfraExplorerWithEnvelope
import fr.sncf.osrd.utils.units.Distance
import fr.sncf.osrd.utils.units.Length
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.meters
import java.lang.Double.isNaN

data class STDCMEdge(
    val timeData: TimeData,
    // Instance used to explore the infra, contains the current
    // underlying edge (block)
    val infraExplorer: InfraExplorerWithEnvelope,
    // Includes this edge's envelope
    val infraExplorerWithNewEnvelope: InfraExplorerWithEnvelope,
    // Node located at the start of this edge
    val previousNode: STDCMNode,
    // Offset of the envelope if it doesn't start at the beginning of the edge
    // This can *not* be used to convert Path / TravelledPath (should reference start of 1st route)
    val envelopeStartOffset: Offset<Block>,
    // True if the edge end is a stop
    val endAtStop: Boolean,
    // Speed at the beginning of the edge
    val beginSpeed: Double,
    // Speed at the end of the edge
    val endSpeed: Double,
    // Edge length
    val length: Length<STDCMEdge>,
    // How long it takes to go from the beginning to the end of the block, taking the
    // standard allowance into account
    val totalTime: Double,
    // If this edge starts at the end of an engineering allowance, this contains
    // some data like its length and extra time. Null if there's no engineering allowance
    // ending here. Overrides any allowance spanning part of the range from previous edges.
    val engineeringAllowance: EngineeringAllowanceData?,
    // Original envelope, not scaled for standard allowance. Needs to be scaled to interpolate times
    // or expected speed.
    val originalEnvelope: Envelope,
) {
    val block = infraExplorer.getCurrentBlock()

    init {
        assert(!isNaN(timeData.earliestReachableTime)) { "STDCM edge starts at NaN time" }
    }

    data class EngineeringAllowanceData(
        /** How long is the allowance section. It always ends at the start of the current edge. */
        val length: Distance,
        /** How much extra time was added on this current allowance range. */
        val extraDuration: Double,
    )

    /** Returns the node at the end of this edge */
    fun getEdgeEnd(graph: STDCMGraph): STDCMNode {
        val previousPlannedNodeRelativeTimeDiff = getPreviousPlannedNodeRelativeTimeDiff()
        val stepTracker = infraExplorer.getStepTracker()
        val newExplorer = infraExplorerWithNewEnvelope.clone()
        newExplorer
            .getStepTracker()
            .moveForward(
                infraExplorer.getCurrentBlock(),
                envelopeStartOffset,
                envelopeStartOffset + length.distance,
            )
        return if (!endAtStop) {
            // We move on to the next block
            STDCMNode(
                timeData.withAddedTime(totalTime, null, null),
                endSpeed,
                newExplorer,
                this,
                null,
                null,
                null,
                previousPlannedNodeRelativeTimeDiff,
                graph.remainingTimeEstimator.invoke(this, null, stepTracker),
                graph,
            )
        } else {
            // New edge on the same block, after a stop
            val nextStop = stepTracker.getStepsInLookahead().first { it.originalStep.stop }
            val stopDuration = nextStop.originalStep.duration
            val locationOnEdge = envelopeStartOffset + length.distance

            STDCMNode(
                timeData.withAddedTime(
                    totalTime,
                    stopDuration,
                    graph.delayManager.getMaxAdditionalStopDuration(
                        infraExplorerWithNewEnvelope,
                        getEdgeEndTime(),
                    ),
                ),
                endSpeed,
                newExplorer,
                this,
                envelopeStartOffset + length.distance,
                stopDuration,
                nextStop.originalStep.plannedTimingData,
                previousPlannedNodeRelativeTimeDiff,
                graph.remainingTimeEstimator.invoke(this, locationOnEdge, stepTracker),
                graph,
            )
        }
    }

    /**
     * Computes the last planned node's previousPlannedNodeRelativeTimeDiff, taking the potentially
     * new total departure time shift into account.
     */
    private fun getPreviousPlannedNodeRelativeTimeDiff(): Double? {
        var currentEdge: STDCMEdge? = this
        while (currentEdge != null) {
            val previousPlannedNode = currentEdge.previousNode
            if (previousPlannedNode.plannedTimingData != null) {
                return previousPlannedNode.getRelativeTimeDiff(timeData)
            }
            currentEdge = previousPlannedNode.previousEdge
        }
        return null
    }

    /**
     * Returns the approximate time of the given offset of the edge. Runs a simple linear
     * interpolation. The updated time data is used to account for any extra departure delay, it
     * should come from the last edge/node.
     */
    fun getApproximateTimeAtLocation(offset: Offset<STDCMEdge>, updatedTimeData: TimeData): Double {
        val updatedEarliestTime = timeData.getUpdatedEarliestReachableTime(updatedTimeData)
        if (length.distance == 0.meters) return updatedEarliestTime // Avoids division by 0
        val offsetRatio = offset.meters / length.meters
        return updatedEarliestTime + (totalTime * offsetRatio)
    }

    override fun toString(): String {
        return "STDCMEdge(timeStart=${timeData.earliestReachableTime}, block=$block)"
    }

    /**
     * Returns the offset on the edge referential from a given block offset, if it's covered by the
     * edge.
     */
    fun edgeOffsetFromBlock(blockOffset: Offset<Block>): Offset<STDCMEdge>? {
        val projectedOffset = Offset<STDCMEdge>(blockOffset - envelopeStartOffset)
        if (projectedOffset.distance < 0.meters || projectedOffset > length) return null
        return projectedOffset
    }

    /** Returns the offset on the block referential from a given edge offset. */
    fun blockOffsetFromEdge(edgeOffset: Offset<STDCMEdge>): Offset<Block> {
        return envelopeStartOffset + edgeOffset.distance
    }

    /** Return the time at which the train quits this edge and moves on to the next. */
    private fun getEdgeEndTime(): Double {
        return timeData.earliestReachableTime + totalTime
    }

    /**
     * Return how much delay we can add on this edge. We need to be given the time added on the next
     * edges (both departure delaying and allowances), as they carry over previous edges.
     */
    fun getMaxAddedDelay(addedAllowanceOnNextEdges: Double): Double {
        val maxAddedDelay = timeData.timeOfNextConflictAtLocation - getEdgeEndTime()
        return maxAddedDelay - addedAllowanceOnNextEdges
    }

    /** Return all delay added to this edge, engineering + departure delay */
    fun getTotalAddedDelayOnEdge(): Double {
        return timeData.delayAddedToLastDeparture + (engineeringAllowance?.extraDuration ?: 0.0)
    }
}
